jobs: TestRegistryLifecycle failed

jobs.TestRegistryLifecycle [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4186913&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4186913&tab=artifacts#/) on master @ [e1068d77afbd39b162978281c9da7cbea49c1c3a](https://github.com/cockroachdb/cockroach/commits/e1068d77afbd39b162978281c9da7cbea49c1c3a):


```
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1282 +0x1514
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1616 +0xfd3
  google.golang.org/grpc.(*Server).serveStreams.func1.2()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0xfd

Goroutine 344 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x615
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:354 +0x6e4
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:753 +0x55a
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).preStart()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1123 +0xbe9
  github.com/cockroachdb/cockroach/pkg/server.(*Server).PreStart()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1926 +0x5084
  github.com/cockroachdb/cockroach/pkg/server.(*Server).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1254 +0x44
  github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/testserver.go:500 +0x64
  github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartServer()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:241 +0x17b
  github.com/cockroachdb/cockroach/pkg/jobs_test.(*registryTestSuite).setUp()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs_test.go:236 +0x69b
  github.com/cockroachdb/cockroach/pkg/jobs_test.TestRegistryLifecycle.func18()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs_test.go:1068 +0x18d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 250 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:919 +0x4d0
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/internal/transport/http2_server.go:594 +0x28fe
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/internal/transport/http2_server.go:639 +0x2fd
  google.golang.org/grpc.(*Server).serveStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:905 +0x289
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:847 +0x64
==================
    jobs_test.go:253: Starting resume
    jobs_test.go:266: Exiting resume
    jobs_test.go:287: Starting FailOrCancel
    jobs_test.go:307: Exiting FailOrCancel
    testing.go:1152: race detected during execution of test
    --- FAIL: TestRegistryLifecycle/dump_traces_on_fail (1.99s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegistryLifecycle.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/util/tracing/crdbspan.go
+++ b/pkg/util/tracing/crdbspan.go
@@ -30,7 +30,10 @@ import (
 // crdbSpan is a span for internal crdb usage. This is used to power SQL session
 // tracing.
 type crdbSpan struct {
+	// tracer is the Tracer that created this span.
 	tracer *Tracer
+	// sp is Span that this crdbSpan is part of.
+	sp *Span
 
 	traceID      tracingpb.TraceID // probabilistically unique
 	spanID       tracingpb.SpanID  // probabilistically unique

--- a/pkg/util/tracing/tracer.go
+++ b/pkg/util/tracing/tracer.go
@@ -418,16 +418,26 @@ func (r *SpanRegistry) getSpanByID(id tracingpb.SpanID) RegistrySpan {
 // VisitSpans calls the visitor callback for every local root span in the
 // registry. Iterations stops when the visitor returns an error. If that error
 // is iterutils.StopIteration(), then VisitSpans() returns nil.
+//
+// The callback should not hold on to the span after it returns.
 func (r *SpanRegistry) VisitSpans(visitor func(span RegistrySpan) error) error {
+	// Take a snapshot of the registry and release the lock.
 	r.mu.Lock()
-	sl := make([]*crdbSpan, 0, len(r.mu.m))
+	spans := make([]spanRef, 0, len(r.mu.m))
 	for _, sp := range r.mu.m {
-		sl = append(sl, sp)
+		// We'll keep the spans alive while we're visiting them below.
+		spans = append(spans, makeSpanRef(sp.sp))
 	}
 	r.mu.Unlock()
 
-	for _, sp := range sl {
-		if err := visitor(sp); err != nil {
+	defer func() {
+		for i := range spans {
+			spans[i].release()
+		}
+	}()
+
+	for _, sp := range spans {
+		if err := visitor(sp.Span.i.crdb); err != nil {
 			if iterutil.Done(err) {
 				return nil
 			}
@@ -543,6 +553,7 @@ func NewTracer() *Tracer {
 			c := &h.crdbSpan
 			*c = crdbSpan{
 				tracer: t,
+				sp:     sp,
 			}
 			sp.i.crdb = c
 			return h
@@ -1354,8 +1365,11 @@ func (t *Tracer) GetActiveSpanByID(spanID tracingpb.SpanID) RegistrySpan {
 	return t.activeSpansRegistry.getSpanByID(spanID)
 }
 
-// VisitSpans invokes the visitor with all active Spans. The function will
-// gracefully exit if the visitor returns iterutil.StopIteration().
+// VisitSpans calls the visitor callback for every local root span in the
+// registry. Iterations stops when the visitor returns an error. If that error
+// is iterutils.StopIteration(), then VisitSpans() returns nil.
+//
+// The callback should not hold on to the span after it returns.
 func (t *Tracer) VisitSpans(visitor func(span RegistrySpan) error) error {
 	return t.activeSpansRegistry.VisitSpans(visitor)
 }

